From 8a7fe7073964bc74a23af734669c9b6309a2ffd1 Mon Sep 17 00:00:00 2001
From: Michal Kubecek <mkubecek@suse.cz>
Date: Mon, 11 Dec 2017 08:51:36 +0100
Subject: [PATCH 21/21] netlink: add netlink handler for sfec (--set-fec)

Implement "ethtool --set-fec <dev>" subcommand using netlink interface
command ETHTOOL_CMD_SET_PARAMS.

This command is not implemented yet in ioctl interface but kernel side
implementation is ready since kernel 4.14-rc1. The command line syntax was
taken from the example in commit message of kernel commit 1a5f3da20bd9
("net: ethtool: add support for forward error correction modes").

Signed-off-by: Michal Kubecek <mkubecek@suse.cz>
---
 ethtool.c        |  3 ++-
 netlink/extapi.h |  1 +
 netlink/params.c | 17 +++++++++++++++++
 3 files changed, 20 insertions(+), 1 deletion(-)

diff --git a/ethtool.c b/ethtool.c
index 126ac605102d..43ece6d2c68a 100644
--- a/ethtool.c
+++ b/ethtool.c
@@ -4933,6 +4933,7 @@ int nl_sring(struct cmd_context *ctx) { return 0; }
 int nl_spause(struct cmd_context *ctx) { return 0; }
 int nl_schannels(struct cmd_context *ctx) { return 0; }
 int nl_seee(struct cmd_context *ctx) { return 0; }
+int nl_sfec(struct cmd_context *ctx) { return 0; }
 #endif
 
 static const struct option {
@@ -5142,7 +5143,7 @@ static const struct option {
 	  "		[ all ]\n"},
 	{ "--show-fec", 1, do_gfec, nl_gfec,
 	  "Show FEC settings"},
-	{ "--set-fec", 1, do_sfec, NULL,
+	{ "--set-fec", 1, do_sfec, nl_sfec,
 	  "Set FEC settings",
 	  "		[ encoding auto|off|rs|baser ]\n"},
 	{ "-h|--help", 0, show_usage, NULL,
diff --git a/netlink/extapi.h b/netlink/extapi.h
index de89e3385e95..d6995886ed70 100644
--- a/netlink/extapi.h
+++ b/netlink/extapi.h
@@ -27,6 +27,7 @@ int nl_sring(struct cmd_context *ctx);
 int nl_spause(struct cmd_context *ctx);
 int nl_schannels(struct cmd_context *ctx);
 int nl_seee(struct cmd_context *ctx);
+int nl_sfec(struct cmd_context *ctx);
 int nl_monitor(struct cmd_context *ctx);
 
 void monitor_usage();
diff --git a/netlink/params.c b/netlink/params.c
index fb2c8670ca96..7b96349f68a1 100644
--- a/netlink/params.c
+++ b/netlink/params.c
@@ -642,3 +642,20 @@ int nl_seee(struct cmd_context *ctx)
 	return nl_set_param(ctx, "--set-eee", seee_params, ETHA_PARAMS_EEE,
 			    ETHA_EEE_MAX);
 }
+
+static const struct param_parser sfec_params[] = {
+	{
+		.arg		= "modes",
+		.type		= ETHA_FEC_MODES,
+		.handler	= nl_parse_bitfield32,
+		.handler_data	= flags_fecenc,
+		.min_argc	= 1,
+	},
+	{}
+};
+
+int nl_sfec(struct cmd_context *ctx)
+{
+	return nl_set_param(ctx, "--set-fec", sfec_params, ETHA_PARAMS_FEC,
+			    ETHA_FEC_MAX);
+}
-- 
2.18.0

